### PR TITLE
Allow external ownership for Modules

### PR DESCRIPTION
**Pull request by Supereg**
# Allow external ownership for Modules

## :recycle: Current situation & Problem
When loading modules dynamically using `loadModule(_:)`, Spezi always takes ownership of the Module. Spezi keeps a strong reference until `unloadModule(_:)` is called. This isn't strictly necessary (with some conditions applying). In some cases it is helpful for framework engineers to not pass ownership to Spezi. Namely, one can make use of being notified when objects are de-initialized (e.g., a framework creating a Module and passing ownership to the framework user and relying on the deinit to be notified when the user deallocated the object).

This PR introduces the concept of `ModuleOwnership` that can be passed to the updated `loadModule(_:ownership:)` method. `external` ownership is available to framework developers via the `APISupport` SPI.
When specifying `external` ownership does not keep any strong references to the loaded Module. This required making some changes how resources are automatically cleared (e.g., `@Dependency`, `@Provide`, `@Modifier` or `@Model` modifiers automatically clean up their resources once their are deallocated). Some features like `EnvironmentAccessible` are not available when using `external` ownership as they would inherently imply strong ownership.
Other modules that depend on such modules either maintain a weak reference if they declare it as an optional dependency (in this case the dependency is automatically removed once the Module de-initializes) or a strong reference if declared as a required dependency. In such a case, ownership is shared with that module for the lifetime of the module that declares a required dependency to an externally managed module.

## :gear: Release Notes 
* Add support for dynamically loading Modules with external ownership. 

## :books: Documentation
Documentation was updated for changed interfaces.

## :white_check_mark: Testing
Additional unit testing was added to verify functionality of dynamically loaded modules with external ownership.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -27,29 +27,41 @@ let package = Package(
     ],
     dependencies: [
         .package(url: "https://github.com/StanfordSpezi/SpeziFoundation", from: "1.0.2"),
-        .package(url: "https://github.com/StanfordBDHG/XCTRuntimeAssertions", from: "1.0.1")
+        .package(url: "https://github.com/StanfordBDHG/XCTRuntimeAssertions", from: "1.0.1"),
+        .package(url: "https://github.com/apple/swift-collections.git", from: "1.1.1"),
+        .package(url: "https://github.com/realm/SwiftLint.git", .upToNextMinor(from: "0.55.1"))
```
Comment (PSchmiedmayer) [inline Package.swift:32]: Wondering if it rather makes sense to add this to the UI Testing app and not the main Swift Package. Building Swift Syntax takes a lot of time and there are also a decent amount of dependencies in SwiftLint that we might not want to pull in any projects relying on Spezi.

Swift collections sounds good to me, scoped and clearly added for the core functionality.

Wondering if we should rather enable the strict concurrency checks as documented in #103.

File: Sources/Spezi/Capabilities/Communication/CollectedModuleValues.swift
Comment (PSchmiedmayer) [inline Sources/Spezi/Capabilities/Communication/CollectedModuleValues.swift:1]: Nice simplification!

File: Sources/Spezi/Dependencies/Property/DependencyContext.swift [hunk 2]
```diff
@@ -52,11 +75,25 @@ class DependencyContext<Dependency: Module>: AnyDependencyContext {
...
 
     func inject(from dependencyManager: DependencyManager) {
-        injectedDependency = dependencyManager.retrieve(optional: isOptional)
+        guard let dependency = dependencyManager.retrieve(module: Dependency.self, optional: isOptional) else {
+            injectedDependency = nil
+            return
+        }
+
+        if isOptional {
+            injectedDependency = .weakDependency(WeaklyStoredModule(dependency))
```
Comment (PSchmiedmayer) [inline Sources/Spezi/Dependencies/Property/DependencyContext.swift:84]: Nice approach with the weak storage!

File: Tests/SpeziTests/DependenciesTests/DependencyTests.swift
Comment (PSchmiedmayer) [inline Tests/SpeziTests/DependenciesTests/DependencyTests.swift:1]: Thank you for the added tests!

**Review by PSchmiedmayer — APPROVED** [review]
Nice work @Supereg; great addition to the overall system design for Spezi!

Only major point I would see is the SwiftLint dependency for the main package. I would propose to rather move that to the UI Testing Application as the main point to develop the package, run all tests, and display linting warnings inside. Let me know what you think and how you want to proceed here.

We might need to wait for SPM to become smarter and not checking out and/or building buildtime or testing dependencies when using a SP as a dependency ...

File: Package.swift (hunk 1 above)
Comment (Supereg) [reply Package.swift:32]: The idea was to integrate SwiftLint directly into Xcode and the development workflow when developing for Spezi frameworks. Like modern IDEs have it since decades 🙂
Smartest move would be to have the build tool rely on a pre-compiled binary (like every other set up does it) and seems like they also acknowledge the issue as for macOS targeted builds they just rely on a pre-built binary (see [here](https://github.com/realm/SwiftLint/blob/3b19e7d99b3a91ed95862659d0c306ca970126f0/Package.swift#L172-L178)). There have also been issues with swiftlint as they tend to specify exact version ranges (e.g., https://github.com/realm/SwiftLint/issues/5533, https://github.com/jpsim/SourceKitten/pull/799) which with the current version was fixed but they already introduced an exact version dependency with the upcoming release.

Your approach of just adding it to the UI Testing app would only lint source files within the testing application, not its dependencies.

And you are right with the build times, just verified this by data current build from this PR took [1m 24s](https://github.com/StanfordSpezi/Spezi/actions/runs/9636148663/job/26573936418?pr=106) and a previous build from the main branch took just [39s](https://github.com/StanfordSpezi/Spezi/actions/runs/9384845089/job/25841560053). 

To summarize, I agree that there are very good points to avoid including swiftlint as a dependency. I think I also tend to not include it for every project.

One approach which could be worth exploring, is including it only if a certain environment variable is set, similar to how firebase controls building their dependencies from source https://github.com/firebase/firebase-ios-sdk/blob/09920d156113a6c32916fd8056ee14b6c05bc434/Package.swift#L1419-L1424. You would need to manually start Xcode from command line once but it would persist as long as this instance is running. The build time would stay low by default with a somewhat easy one shot command to enter "developer mode" for a day (or as long Xcode is running).

Comment (Supereg) [reply Package.swift:32]: Added Strict Concurrency checking with the latest commit 👍 
Further I introduced the `SPEZI_DEVELOPMENT_SWIFTLINT` environment variable that conditionally allows to define swiftlint as a dependency.

Comment (Supereg) [reply Package.swift]: This PR technically supersedes #104 now. This PR additionally also fixes the warnings. However, we currently do not mark warnings as errors.

Comment (PSchmiedmayer) [reply Package.swift]: 100% agree; a quite unfortunate state at this point 🤷‍♂️

I like the idea of having a environment variable to address this issue and allowing us to easily enable SwiftLint conditionally within an dev environment 👍

Thanks for fixing this here @Supereg! We should then use #104 for marking warnings as errors @pauljohanneskraft, that is something that would be great independent of the exact warnings fixed here and would be great as an addition to the overall build pipeline.

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for the improvements; looks great!
